src/models/blockchain/transaction/primitive_witness.rs

```diff
@@ -410,6 +428,111 @@ impl PrimitiveWitness {
+        let new_addition_records = mutator_set_update.additions;
+        let mut new_removal_records = mutator_set_update.removals;
+
+        new_removal_records.reverse();
+        let mut block_removal_records: Vec<&mut RemovalRecord> =
+            new_removal_records.iter_mut().collect::<Vec<_>>();
+        let mut msa_state: MutatorSetAccumulator = self.mutator_set_accumulator.clone();
+        let mut transaction_removal_records: Vec<RemovalRecord> = self.kernel.inputs.clone();
+        let mut transaction_removal_records: Vec<&mut RemovalRecord> =
+            transaction_removal_records.iter_mut().collect();
+
+        let mut primitive_witness = self;
+
+        // Apply all addition records in the block
+        for block_addition_record in new_addition_records {
+            // Batch update block's removal records to keep them valid after next addition
+            RemovalRecord::batch_update_from_addition(&mut block_removal_records, &msa_state);
+
+            // Batch update transaction's removal records
+            RemovalRecord::batch_update_from_addition(&mut transaction_removal_records, &msa_state);
+
+            // Batch update primitive witness membership proofs
+            let membership_proofs = &mut primitive_witness
+                .input_membership_proofs
+                .iter_mut()
+                .collect_vec();
+            let own_items = primitive_witness
+                .input_utxos
+                .utxos
+                .iter()
+                .map(Tip5::hash)
+                .collect_vec();
+            MsMembershipProof::batch_update_from_addition(
+                membership_proofs,
+                &own_items,
+                &msa_state,
+                &block_addition_record,
+            )
+            .expect("MS MP update from add must succeed in wallet handler");
+
+            msa_state.add(&block_addition_record);
+        }
+
+        while let Some(removal_record) = block_removal_records.pop() {
+            // Batch update block's removal records to keep them valid after next removal
+            RemovalRecord::batch_update_from_remove(&mut block_removal_records, removal_record);
+
+            // batch update transaction's removal records
+            // Batch update block's removal records to keep them valid after next removal
+            RemovalRecord::batch_update_from_remove(
+                &mut transaction_removal_records,
+                removal_record,
+            );
+
+            // Batch update primitive witness membership proofs
+            let membership_proofs = &mut primitive_witness
+                .input_membership_proofs
+                .iter_mut()
+                .collect_vec();
+
+            MsMembershipProof::batch_update_from_remove(membership_proofs, removal_record)
+                .expect("MS MP update from add must succeed in wallet handler");
+
+            msa_state.remove(removal_record);
+        }
```

Comment:
I've seen this logic before. Are you sure this not repeated code? If so, I would strongly suggest to factor out a function for it (if it doesn't exist already).